workflow/controller/taskresult.go

```diff
@@ -86,6 +86,8 @@ func (woc *wfOperationCtx) taskResultReconciliation() {
 			if old.Outputs != nil && newNode.Outputs.ExitCode == nil { // prevent overwriting of ExitCode
 				newNode.Outputs.ExitCode = old.Outputs.ExitCode
 			}
+			// Add outputs to global scope here to ensure that they are reflected in archive.
+			woc.addOutputsToGlobalScope(newNode.Outputs)
```

Comment:
I see. So, basically you moved the propagation to `WorkflowStatus.Outputs` into `taskResultReconciliation()` so you don't need to have a delay anymore, right? I think that seems good.

We should just make sure that there's no logic between here and `podReconciliation()` that assumes that it hasn't been set or something (probably not but would you mind verifying that?).

Reply:
A nice side effect is that it removes the delay, yes, but It's not just for removing the delay.

It fixes the condition where a node status has been set to its final state, but the task result wasn't reconciled on the last pass through the operator, so the outputs aren't reflected in the global scope (and therefore don't show up in the API). This can happen, for example, when the workflow is stopped. I reference the rationale behind this [here](https://github.com/argoproj/argo-workflows/pull/12402#issuecomment-1871710821). Notably, `podReconciliation()` checks if there has been a state change for a node before calling `woc.addOutputsToGlobalScope()`. If the node is marked as `wfv1.NodeFailed`, for example, but reconciliation hasn't completed for the node's outputs yet, then on the next pass through `podReconciliation()`, `newState` will be nil, and `woc.addOutputsToGlobalScope(newState.Outputs)` won't run.

```
	if newState := woc.assessNodeStatus(pod, node); newState != nil {
		woc.addOutputsToGlobalScope(newState.Outputs)
```

**Example flow:**

Stop workflow -> running node marked failed (outputs haven't been written by wait container yet) -> newState (wfv1.NodeFailed) returned by `assessNodeStatus(pod, node)` -> `addOutputsToGlobalScope()` called inside `podReconciliation`, but outputs still not reconciled, so they are not included in global scope -> operator continues looping until `woc.checkReconciliationComplete() && !woc.hasDaemonNodes()` resolves true, and `common.LabelKeyCompleted` is set. -> wait container finishes writing outputs -> `woc.taskResultReconciliation()` deals with them and makes its own call to `woc.addOutputsToGlobalScope(newNode.Outputs)` (remember, this wouldn't get called in `podReconciliation()` because there is no new node state) -> Outputs reconciled, therefore allowing `common.LabelKeyCompleted` to be set to true.

The only other references to Status.Outputs are in `addOutputsToLocalScope` inside `steps.go` and `dag.go` and in the `AddParamToGlobalScope` util, so there is no dependency on Status.Outputs being nil, as far as I can see.

Obviously it's also referenced in `get.go` for the API.

Reply:
thanks for checking